### ISSUE TITLE
[Message] Create a Message component

### ISSUE THREAD

**Issue by sroze** [symfony/symfony#24326]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | 4.1

# Context
Every day, more applications require messages going out of the inner context of the application. This could be because of a micro-service architecture, 3rd party dependencies or asynchronous processing. I believe we should have all the tools within Symfony to fulfil these needs.

# Current attempts
#23650 offers to add an asynchronous processing to the EventDispatcher. Everybody more or less agrees that this isn't EventDispatcher's responsibility and this should be its own component.

#23315 offer to create two AMQP and Worker components. My personal point of view is that there are no real reason of being two components and that they are too specific. Also, we should leverage the experience being built within Swarrot or php-enqueue.

# A Message component
The philosophy behind a Message component is that we can generalise a lot of use cases behind a few concepts:

- **Message**
The user’s domain object. This won't require any specific interface.

- **MessageBus**
Middleware powered, it will be responsible of calling the correct message handlers. The middlewares adds an extension point that will be used to enforce rules by the user or libraries (for example, transactional consistency for commands within a CQRS application)

- **MessageHandler**
Will be called to handle the message. It can be one or more objects registered to handle the message.
 
- **MessageConsumer**
This will be responsible of getting the messages from a source (queuing system, command-line, HTTP interface, …) and dispatching it to the message bus.

- **MessageProducer**
Will be responsible of creating the message and sending it to a target (queuing system, task management system, HTTP interface, …) 

- **MessageEncoder** and **MessageDecoder**
A way to encode and decode the user-land messages into something that is serializable/deserializable from the producers and consumers. The API should be something like `encode($message) : array` and `decode(array $encodedMessage) : mixed` I believe, to have a single representation (as an array). 

The value in such mechanism is that this messaging component is generalised to allow the communication from/to applications (through an asynchronous queuing system or not).

# The consumer abstraction

One of my favourite outcome of such component is the consumer abstraction. As a user, I will send a message to the bus and this message will either be handled by a "handler" directly within my application or sent to a "producer". This allows a smooth progression within the applications' architecture, enabling beginners to easily plug a queuing mechanism. Here is how I envisage the flow:

1. User sends a message to the message bus
2. A middleware knows how to get the handlers. If there are handlers for this message within the application, it will call them. If no handler, call the next middeware
3. This next middleware knows how to get the correct producer for the message (could be a default one, and configuration powered). It will send the message to the middleware.

# What will this component contain?

1. A MessageBus with 2 default middlewares:
    1. `CallMessageHandlersMiddleware`
    2. `SendMessageToProducersMiddleware`
2. An AMQP implementation of a consumer and producer based on `php-amqp`, not more. The other implementation should IMHO be in [Swarrot](https://github.com/swarrot) and/or [php-enqueue](https://github.com/php-enqueue).
3. A `MessageBundle` bundle that will gives the following capabilities:
    1. Register a handler with a `message_handler` tag
    2. A `message:consume` command that will start a consumer
4. (Optional) A DataCollector to display the various sent/handled messages within the WDT and Profiler.

# Usage

## Without Symfony Bundle

```php
$bus = new MessageBus([
    new CallMessageHandlersMiddleware([
        UserCreatedMessage::class => [new UserCreatedMessageHandler(/* ... */),],
    ]),
    new SendMessageToProducersMiddleware([
        UserMessageToBeSent::class => [new AMQPMessageProducer(/* ... */),],
    ]),
]);

$bus->handle(new UserCreatedMessage('...'));
```

## Within Symfony

```yaml
# framework.yaml
framework:
    message_bus:
        message_producer_mapping:
            App\Message\UserMessageToBeSent: swarrot.publisher
```

```yaml
# services.yaml
services:
    App\Handler\UserCreatedMessageHandler:
        tags: [{name: message_handler, handles: App\Message\UserCreatedMessage}]
```

---

I'd like to have a few weeks of discussions on the idea before starting to move the code I have from our [message library](https://github.com/continuouspipe/message) and [SimpleBus' MessageBus](https://github.com/SimpleBus/MessageBus) within a Component PR. Please feel free to challenge as much as possible so we have the best component for Symfony :)

**Comment by javiereguiluz** [comment]
To have more context and better analyze this proposal, could you please share some links to the implementation of this feature in relevant programming languages (Java, C#, Go) and popular web frameworks and technologies (Rails, Django, Node.js). Thanks!

**Comment by sstok** [comment]
FYI https://github.com/prooph/ is very popular for this purpose but the Symfony integration is in desperate need of updating (it's currently not compatible with Symfony 4).

I could actually life a simpler implementation then Prooph because I only use the ServiceBus, but SimpleBus misses some parts I need (like separate buses per context).

**Comment by mvrhov** [comment]
Pulling simplebus in and adding missing features would probably be best :) Like return value from command so even query buss could be built

**Comment by iltar** [comment]
https://github.com/thephpleague/tactician-bundle is a pretty good package from what I can tell

**Comment by sroze** [comment]
@javiereguiluz here is the comparison with the two buses I've had a look at.

# Java Message Service

One very famous example is JMS. Framework like Spring are using JMS to provide a simple way to exchange messages, by [providing annotations to define listeners](https://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#jms-annotated) consumers and easiest ways to [create JMS messages](https://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#jms-sending).

The userland messages (that do not require any specific implementation details) are encapsulated into an [internal `Message` object](https://docs.oracle.com/javaee/7/api/javax/jms/Message.html), that the `MessageProducer`s know how to send to a *destination*. The *destinations* is a concept that defines where the message will be sent, that's how the routing is made.

To compare with this proposal, the destination mechanism is handled here by the producer mapping (i.e. this message go to these producers). Message encoding/decoding will be handled by the `MessageEncoder` and `MessageDecoder`, which take an object and will return an `array` to be sent. The user-land objects do not have any specifics, same with JMS.

# Laravel

Laravel comes with [a Command Bus](https://laravel.com/docs/5.0/bus), mostly focused on commands as the name reflects it. The bus has a `dispatch` method, that is called using an object that extends an abstract `Command` class. This bus has what they call a ["command pipeline"](https://laravel.com/docs/5.0/bus#command-pipeline) which allows to use middlewares to wrap the commands within a transaction or similar operations.

On top of this bus, there is a ["Queue" service](https://laravel.com/docs/5.0/queues#basic-usage) that allows to define various connections to queue vendors and send this message to the queue, if your message implement a specific [interface `ShouldQueue`](https://laravel.com/api/5.3/Illuminate/Contracts/Queue/ShouldQueue.html). 

The routing is made by manually setting the queue name while sending the message (or defaulting to the single connection). On the UX perspective, there is a [nice helper which transforms an HTTP request to a command](https://laravel.com/docs/5.0/bus#dispatching-commands), this could be done with an HTTP consumer.

**Comment by mvrhov** [comment]
Both simplebus and tactician are simple against prooph.
The + (against tactician) of simplebus is that you can trigger/record events during command execution. The - is that you cannot build query bus because you can't return from it.

**Comment by makasim** [comment]
@sroze FYI [queue-interop](https://github.com/queue-interop/queue-interop) is a heavily inspired by Java JMS and follows their spec.

**Comment by javiereguiluz** [comment]
Thank you all for the references! It's always good to know how others have solved the same problem.

**Comment by makasim** [comment]
Enqueue Client  (a high-level MQ abstraction) covers all parts described in the description.  Here are Enqueue analogs for proposed clasess and interfaces. 

* Message - https://github.com/php-enqueue/enqueue-dev/blob/master/pkg/enqueue/Client/Message.php. A producer can work with it but it is not required. It is possible to use primitives, an array of primitives, something serializable to json.

MessageBus - https://github.com/php-enqueue/enqueue-dev/blob/master/pkg/enqueue/Client/DriverInterface.php

MessageHandler - https://github.com/queue-interop/queue-interop/blob/master/src/PsrProcessor.php

MessageConsumer - https://github.com/php-enqueue/enqueue-dev/blob/master/pkg/enqueue/Consumption/QueueConsumer.php

MessageProducer - https://github.com/php-enqueue/enqueue-dev/blob/master/pkg/enqueue/Client/ProducerInterface.php.

MessageEncoder and MessageDecoder - by default Enqueue encodes everything to JSON. though it could be changed in an extension.

**Comment by sroze** [comment]
> FYI queue-interop is a heavily inspired by Java JMS and follows their spec.

@makasim I think it would be great to have a [queue-interop](https://github.com/queue-interop/queue-interop) adapter (i.e. a `MessageProducer` and `MessageConsumer` implementation), so it can leverage all the work already done on the transport implementations.

> The + (against tactician) of simplebus is that you can trigger/record events during command execution. The - is that you cannot build query bus because you can't return from it.

I did not explicitly mentioned it here, but you're right, I think it's important to be able to get the return value from the bus, if any. Obviously, when the message is sent to an asynchronous producer, you won't be able to get the result of your operations. I also think that tactician is a really good base of inspiration to create such component, but I'd like it to be more than a command bus (i.e. also focusing on messages - could be commands, could be queries, operational things, ...) across applications through queuing systems but also HTTP APIs for example.

**Comment by andersonamuller** [comment]
This book is a good reference for the patterns described here http://www.enterpriseintegrationpatterns.com/patterns/messaging/

**Comment by dunglas** [comment]
I like this proposal! Having adapters for Google Pub/Sub and Amazon SQS would be great.

**Comment by makasim** [comment]
@dunglas enqueue already provides [Amazon SQS](https://github.com/php-enqueue/enqueue-dev/blob/master/docs/transport/sqs.md) and [Google Pub\Sub](https://github.com/php-enqueue/enqueue-dev/blob/master/docs/transport/gps.md) transports. They are [queue interop](https://github.com/queue-interop/queue-interop) compatible.

**Comment by sroze** [comment]
Great, looks like enqueue could be the first adapter to start with :)

**Comment by andreybolonin** [comment]
+1

**Comment by sroze** [comment]
@andreybolonin FYI, everything is happening in the pull-request now,  see https://github.com/symfony/symfony/pull/24411 :)
